adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -504,15 +511,15 @@ protected final void onSubscribe(
                     // we do not support subscribing to commands using QoS 2
                     result = Future.failedFuture(new IllegalArgumentException("QoS 2 not supported for command subscription"));
                 } else {
-                    result = createCommandConsumer(endpoint, cmdSub).map(consumer -> {
+                    result = createCommandConsumer(endpoint, cmdSub, cmdHandler).map(consumer -> {
                         // make sure that we close the consumer and notify downstream
                         // applications when the connection is closed
                         endpoint.closeHandler(c -> {
                             // do not use the current span for sending the disconnected event
                             // because that span will (usually) be finished long before the
                             // connection is closed
                             final Span closeHandlerSpan = newSpan("Send Disconnected Event", endpoint, authenticatedDevice);
-                            sendDisconnectedTtdEvent(cmdSub.getTenant(), cmdSub.getDeviceId(), authenticatedDevice, closeHandlerSpan.context()) 
+                            sendDisconnectedTtdEvent(cmdSub.getTenant(), cmdSub.getDeviceId(), authenticatedDevice, closeHandlerSpan.context())
```

Comment:
when the client closes the connection we need to clean up the state in the `CommandHandler`, in particular, we need to close all consumers of all of the client's subscriptions and remove the subscriptions. I believe that any pending PUBACKs will be cleaned up by the timers, right?

Reply:
Inside `protected final void close(final MqttEndpoint endpoint, final Device authenticatedDevice)` the consumers can be closed and entries can be removed. Yes PUBACKs will be cleaned by the timers.

Reply:
@sophokles73 In `onSubscribe` method the `closeHandler ` as given in the below code is being registered. I see consumer is being closed here but it is not right when there are more than one subscriptions.  IMHO this part `endpoint.closeHandler(` should be removed. I will close the consumers for all subscriptions within the `close` method and remove the subscription entries from `CommandHandler`. 
```
                        endpoint.closeHandler(c -> {
                            // do not use the current span for sending the disconnected event
                            // because that span will (usually) be finished long before the
                            // connection is closed
                            final Span closeHandlerSpan = newSpan("Send Disconnected Event", endpoint, authenticatedDevice);
                            sendDisconnectedTtdEvent(cmdSub.getTenant(), cmdSub.getDeviceId(), authenticatedDevice, closeHandlerSpan.context())
                            .setHandler(sendAttempt -> {
                                consumer.close(null);
                                close(endpoint, authenticatedDevice, cmdHandler);
                                closeHandlerSpan.finish();
                            });
                        });
```